File: lineread.py
"""
Raw-mode line reader with correct multi-line paste handling.

WHY THIS EXISTS
---------------
`input()` plus libedit cannot read a pasted paragraph. The tty is in canonical
mode, so it hands over a line as soon as it sees a newline: pasting

    Reply to this note:
    The meeting moved to 3pm.

auto-submits "Reply to this note:" before you touch Enter, and the rest becomes
separate turns. Draining the leftover lines from stdin recovers most of it, but
not a paste whose selection has no trailing newline -- and that is the common
case when copying a paragraph. The final partial line stays wedged in the tty's
line buffer; measured, it is NOT retrievable by switching to cbreak and reading.

The only real fix is to stop the tty from buffering lines. In raw mode every
byte arrives immediately, so a paste lands as one burst that we can assemble
ourselves. Two independent mechanisms decide what is a paste:

  * BRACKETED PASTE. We ask the terminal to wrap pastes in \\e[200~ ... \\e[201~
    (supported by Terminal.app, iTerm2, VS Code, tmux). Everything between the
    markers is literal text, newlines included.
  * BURST HEURISTIC, for terminals that lack it. On a newline, if more input is
    already waiting, it came from a paste rather than a keystroke -- no human
    produces the next byte within 10 ms of pressing Return.

Editing is deliberately modest: characters, backspace, left/right, home/end, and
in-memory history, plus Ctrl-J to insert a newline. Once the buffer contains a newline (i.e. a paste happened),
history navigation is disabled, because redrawing a wrapped multi-line buffer
correctly is a much bigger job than this tool needs -- after pasting a paragraph
you press Return.

Nothing is persisted. History dies with the process.
"""

import codecs
import select
import sys

try:
    import termios
    import tty
    _HAVE_TTY = True
except ImportError:                      # non-POSIX
    _HAVE_TTY = False

PASTE_START, PASTE_END = "\x1b[200~", "\x1b[201~"
BURST_GAP = 0.01                         # a keystroke never follows this fast


def available() -> bool:
    return _HAVE_TTY and sys.stdin.isatty() and sys.stdout.isatty()


class LineReader:
    def __init__(self):
        self.history = []                # in memory only, never written
        self._dec = codecs.getincrementaldecoder("utf-8")("replace")

    # ------------------------------------------------------------------
    def read(self, prompt: str) -> str:
        """Read one message. Raises EOFError on Ctrl-D, KeyboardInterrupt on Ctrl-C."""
        if not available():              # piped stdin, or no tty: plain readline
            line = sys.stdin.readline()
            if not line:
                raise EOFError
            return line.rstrip("\n")

        fd = sys.stdin.fileno()
        old = termios.tcgetattr(fd)
        sys.stdout.write(prompt + PASTE_ON)
        sys.stdout.flush()
        try:
            # TCSANOW, not tty.setraw's default TCSAFLUSH: TCSAFLUSH DISCARDS
            # pending input, which would silently eat a paste that arrives while
            # the prompt is being drawn, and it blocks until output drains.
            tty.setraw(fd, termios.TCSANOW)
            return self._loop(prompt)
        finally:
            termios.tcsetattr(fd, termios.TCSADRAIN, old)
            sys.stdout.write(PASTE_OFF)
            sys.stdout.flush()

    # ------------------------------------------------------------------
    def _more_pending(self, gap=BURST_GAP) -> bool:
        return bool(select.select([sys.stdin], [], [], gap)[0])

    def _getch(self) -> str:
        """One decoded character (blocking), handling split UTF-8 sequences."""
        while True:
            b = sys.stdin.buffer.raw.read(1)
            if not b:
                raise EOFError
            s = self._dec.decode(b)
            if s:
                return s

    def _redraw(self, prompt, buf, cur):
        """Repaint the current visual line -- the text after the last newline."""
        head, _, tail = buf.rpartition("\n")
        prefix = "" if head or buf.endswith("\n") else prompt
        line_start = len(buf) - len(tail)
        sys.stdout.write("\r" + prefix + tail + "\x1b[K")
        back = len(tail) - (cur - line_start)
        if back > 0:
            sys.stdout.write(f"\x1b[{back}D")
        sys.stdout.flush()

    def _loop(self, prompt):
        buf, cur, hidx = "", 0, len(self.history)
        pending = ""                      # partial escape sequence
        skip_lf = False                   # just consumed the CR of a CRLF pair

        while True:
            ch = self._getch()

            # ---- escape sequences: paste markers, arrows ----
            if ch == "\x1b" or pending:
                pending += ch if not pending else ch
                if pending == "\x1b":
                    continue
                if PASTE_START.startswith(pending):
                    if pending == PASTE_START:
                        text = self._read_paste()
                        buf = buf[:cur] + text + buf[cur:]
                        cur += len(text)
                        # Echo only up to the final newline; _redraw owns the
                        # current visual line. Echoing the whole paste AND then
                        # redrawing prints the last line twice.
                        if "\n" in text:
                            head = text[: text.rindex("\n") + 1]
                            sys.stdout.write(head.replace("\n", "\r\n"))
                        self._redraw(prompt, buf, cur)
                        pending = ""
                    continue
                if pending in ("\x1b[", "\x1bO"):
                    continue
                key, pending = pending, ""
                if key.endswith("D") and cur > 0:            # left
                    cur -= 1
                elif key.endswith("C") and cur < len(buf):   # right
                    cur += 1
                elif key.endswith("A") and "\n" not in buf:  # up: history
                    if hidx > 0:
                        hidx -= 1
                        buf = self.history[hidx]
                        cur = len(buf)
                elif key.endswith("B") and "\n" not in buf:  # down
                    if hidx < len(self.history) - 1:
                        hidx += 1
                        buf = self.history[hidx]
                        cur = len(buf)
                    else:
                        hidx, buf, cur = len(self.history), "", 0
                elif key.endswith("H"):
                    cur = 0
                elif key.endswith("F"):
                    cur = len(buf)
                self._redraw(prompt, buf, cur)
                continue

            # ---- newline handling ----
            # Return arrives as CR (raw mode disables ICRNL). A bare LF therefore
            # only comes from a paste or from Ctrl-J, so LF always inserts and
            # only CR can submit. CRLF pastes are collapsed via skip_lf.
            if ch == "\n":
                if skip_lf:
                    skip_lf = False
                    continue
                buf = buf[:cur] + "\n" + buf[cur:]
                cur += 1
                sys.stdout.write("\r\n")
                sys.stdout.flush()
                continue
            if ch == "\r":
                if self._more_pending():      # more bytes behind it => a paste
                    buf = buf[:cur] + "\n" + buf[cur:]
                    cur += 1
                    skip_lf = True            # swallow a following LF
                    sys.stdout.write("\r\n")
                    sys.stdout.flush()
                    continue
                sys.stdout.write("\r\n")
                sys.stdout.flush()
                if buf.strip():
                    self.history.append(buf)
                return buf
            if ch == "\x03":                                  # Ctrl-C
                sys.stdout.write("\r\n")
                raise KeyboardInterrupt
            if ch == "\x04":                                  # Ctrl-D
                if not buf:
                    sys.stdout.write("\r\n")
                    raise EOFError
                continue
            if ch in ("\x7f", "\b"):                          # backspace
                if cur > 0:
                    crossed = buf[cur - 1] == "\n"
                    buf = buf[:cur - 1] + buf[cur:]
                    cur -= 1
                    if crossed:            # cannot repaint upward; start clean
                        sys.stdout.write("\x1b[A\r\x1b[K")
                self._redraw(prompt, buf, cur)
                continue
            if ch == "\x15":                                  # Ctrl-U: clear line
                head, sep, _ = buf.rpartition("\n")
                buf = head + sep
                cur = len(buf)
                self._redraw(prompt, buf, cur)
                continue
            if ch == "\x01":                                  # Ctrl-A
                cur = 0
                self._redraw(prompt, buf, cur)
                continue
            if ch == "\x05":                                  # Ctrl-E
                cur = len(buf)
                self._redraw(prompt, buf, cur)
                continue
            if ch < " " and ch not in ("\t",):
                continue                                      # ignore other controls

            skip_lf = False
            buf = buf[:cur] + ch + buf[cur:]
            cur += 1
            if cur == len(buf) and self._more_pending(0):
                # Mid-burst (a paste on a terminal without bracketed paste):
                # append-only echo. A full redraw per character is O(n^2) output
                # and visibly flickers. The last character of the burst falls
                # through to the real redraw below.
                sys.stdout.write(ch)
            else:
                self._redraw(prompt, buf, cur)

    def _read_paste(self) -> str:
        """Read up to the bracketed-paste end marker."""
        out = ""
        while not out.endswith(PASTE_END):
            out += self._getch()
        return out[: -len(PASTE_END)].replace("\r\n", "\n").replace("\r", "\n")


PASTE_ON, PASTE_OFF = "\x1b[?2004h", "\x1b[?2004l"


# ----------------------------------------------------------------------


def _selftest():
    """Drive a real pty: typing, editing, history, and both paste paths."""
    import os
    import pty
    import time

    script = (
        "import sys;sys.path.insert(0,%r)\n"
        "from lineread import LineReader\n"
        "r = LineReader()\n"
        "for _ in range(7):\n"
        "    try: m = r.read('> ')\n"
        "    except EOFError: print('EOF');break\n"
        "    print('MSG=' + repr(m))\n"
        "    sys.stdout.flush()\n"
    ) % os.path.dirname(os.path.abspath(__file__))
    open("/tmp/_lr_probe.py", "w").write(script)

    # Each case is a list of byte chunks sent with a gap between them, which is
    # what really happens: the paste arrives as one burst, then the human presses
    # Return a moment later. Pasting never submits on its own.
    cases = [
        ("typed + backspace",      [b"helloX\x7f", b"\r"],            "'hello'"),
        ("paste, trailing NL",     [b"aaa\nbbb\nccc\n", b"\r"],       "'aaa\\nbbb\\nccc\\n'"),
        ("paste, NO trailing NL",  [b"aaa\nbbb\nccc", b"\r"],         "'aaa\\nbbb\\nccc'"),
        ("bracketed paste",        [b"\x1b[200~one\ntwo\x1b[201~", b"\r"], "'one\\ntwo'"),
        ("CRLF paste",             [b"win\r\ndows\r\n", b"\r"],       "'win\\ndows\\n'"),
        ("Ctrl-J newline",         [b"one", b"\n", b"two", b"\r"],    "'one\\ntwo'"),
        ("history recall (up)",    [b"\x1b[A", b"\r"],                "'one\\ntwo'"),
    ]
    pid, fd = pty.fork()
    if pid == 0:
        os.execvp("python3", ["python3", "/tmp/_lr_probe.py"])

    # A pty master must be drained continuously: if its output buffer fills, the
    # child blocks inside tcsetattr/write and the test deadlocks.
    import threading
    seen, lock, stop = [], threading.Lock(), []

    def drain():
        while not stop:
            try:
                if select.select([fd], [], [], 0.05)[0]:
                    c = os.read(fd, 4096)
                    if not c:
                        break
                    with lock:
                        seen.append(c.decode(errors="replace"))
            except OSError:
                break

    t = threading.Thread(target=drain, daemon=True)
    t.start()
    time.sleep(1.0)
    ok = True
    for label, chunks, want in cases:
        with lock:
            seen.clear()
        for c in chunks:
            os.write(fd, c)
            time.sleep(0.35)          # gap: a burst, then a keystroke
        time.sleep(0.6)
        with lock:
            chunk = "".join(seen)
        import re as _re
        got = _re.findall(r"MSG=(.*)", chunk)
        good = bool(got) and got[-1].strip() == want
        ok &= good
        print(f"  [{'ok' if good else 'FAIL'}] {label:28s} -> "
              f"{got[-1].strip() if got else '(nothing)'}")
        if not good:
            print(f"        wanted {want}\n        raw: {chunk[:300]!r}")
    stop.append(1)
    try:
        os.close(fd)
    except OSError:
        pass
    print("\nALL PASS" if ok else "\nFAILURES ABOVE")
    return ok


if __name__ == "__main__":
    raise SystemExit(0 if _selftest() else 1)

File: cpp/src/lineread.cpp
#include "lineread.hpp"

#include <sys/select.h>
#include <termios.h>
#include <unistd.h>

#include <cerrno>
#include <cstdio>
#include <iostream>

#include "unicode.hpp"

namespace lineread {
namespace {

const char* kPasteStart = "\x1b[200~";
const char* kPasteEnd = "\x1b[201~";
const char* kPasteOn = "\x1b[?2004h";
const char* kPasteOff = "\x1b[?2004l";

// A keystroke never follows this fast, so anything already waiting is a paste.
constexpr double kBurstGap = 0.01;

void out(const std::string& s) {
  fwrite(s.data(), 1, s.size(), stdout);
  fflush(stdout);
}

bool more_pending(double gap) {
  fd_set fds;
  FD_ZERO(&fds);
  FD_SET(STDIN_FILENO, &fds);
  struct timeval tv;
  tv.tv_sec = static_cast<time_t>(gap);
  tv.tv_usec = static_cast<suseconds_t>((gap - tv.tv_sec) * 1e6);
  return select(STDIN_FILENO + 1, &fds, nullptr, nullptr, &tv) > 0;
}

// One raw byte. Returns false at EOF.
bool getbyte(unsigned char& b) {
  ssize_t n;
  do {
    n = ::read(STDIN_FILENO, &b, 1);
  } while (n < 0 && errno == EINTR);
  return n == 1;
}

size_t count_cp(const std::string& s, size_t from, size_t to) {
  size_t n = 0;
  for (size_t i = from; i < to && i < s.size();) {
    i += static_cast<size_t>(uni::seq_len(static_cast<unsigned char>(s[i])));
    ++n;
  }
  return n;
}

size_t prev_cp(const std::string& s, size_t cur) {
  if (cur == 0) return 0;
  size_t i = cur - 1;
  // step back over continuation bytes (10xxxxxx)
  while (i > 0 && (static_cast<unsigned char>(s[i]) & 0xC0) == 0x80) --i;
  return i;
}

size_t next_cp(const std::string& s, size_t cur) {
  if (cur >= s.size()) return s.size();
  size_t n = static_cast<size_t>(uni::seq_len(static_cast<unsigned char>(s[cur])));
  return std::min(cur + n, s.size());
}

std::string crlf(const std::string& s) {
  std::string r;
  for (char c : s) {
    if (c == '\n') r += "\r\n";
    else r += c;
  }
  return r;
}

}  // namespace

bool available() { return isatty(STDIN_FILENO) && isatty(STDOUT_FILENO); }

namespace {

// Repaint the current visual line -- the text after the last newline.
void redraw(const std::string& prompt, const std::string& buf, size_t cur) {
  size_t nl = buf.rfind('\n');
  size_t line_start = (nl == std::string::npos) ? 0 : nl + 1;
  std::string tail = buf.substr(line_start);
  // The prompt only belongs on the first visual line.
  const std::string prefix = (nl == std::string::npos) ? prompt : "";

  std::string s = "\r" + prefix + tail + "\x1b[K";
  size_t back = count_cp(buf, cur, buf.size());
  if (back > 0) s += "\x1b[" + std::to_string(back) + "D";
  out(s);
}

std::string read_paste() {
  // Read up to the bracketed-paste end marker.
  std::string acc;
  const std::string end = kPasteEnd;
  unsigned char b;
  while (acc.size() < end.size() ||
         acc.compare(acc.size() - end.size(), end.size(), end) != 0) {
    if (!getbyte(b)) break;
    acc.push_back(static_cast<char>(b));
  }
  if (acc.size() >= end.size() &&
      acc.compare(acc.size() - end.size(), end.size(), end) == 0) {
    acc.resize(acc.size() - end.size());
  }
  // Normalize line endings the way the terminal may have mangled them.
  std::string r;
  for (size_t i = 0; i < acc.size(); ++i) {
    if (acc[i] == '\r') {
      if (i + 1 < acc.size() && acc[i + 1] == '\n') ++i;
      r += '\n';
    } else {
      r += acc[i];
    }
  }
  return r;
}

Line loop(const std::string& prompt, std::vector<std::string>& history) {
  std::string buf;
  size_t cur = 0;
  size_t hidx = history.size();
  std::string pending;   // partial escape sequence
  bool skip_lf = false;  // just consumed the CR of a CRLF pair

  const std::string paste_start = kPasteStart;

  while (true) {
    unsigned char b;
    if (!getbyte(b)) return {Status::Eof, ""};
    char ch = static_cast<char>(b);

    // ---- escape sequences: paste markers, arrows ----
    if (ch == '\x1b' || !pending.empty()) {
      pending += ch;
      if (pending == "\x1b") continue;
      if (paste_start.compare(0, pending.size(), pending) == 0) {
        if (pending == paste_start) {
          std::string text = read_paste();
          buf.insert(cur, text);
          cur += text.size();
          // Echo only up to the final newline; redraw() owns the current visual
          // line. Echoing the whole paste AND then redrawing prints the last
          // line twice.
          size_t nl = text.rfind('\n');
          if (nl != std::string::npos) out(crlf(text.substr(0, nl + 1)));
          redraw(prompt, buf, cur);
          pending.clear();
        }
        continue;
      }
      if (pending == "\x1b[" || pending == "\x1bO") continue;

      const char key = pending.back();
      pending.clear();
      if (key == 'D' && cur > 0) {              // left
        cur = prev_cp(buf, cur);
      } else if (key == 'C' && cur < buf.size()) {  // right
        cur = next_cp(buf, cur);
      } else if (key == 'A' && buf.find('\n') == std::string::npos) {  // up
        if (hidx > 0) {
          --hidx;
          buf = history[hidx];
          cur = buf.size();
        }
      } else if (key == 'B' && buf.find('\n') == std::string::npos) {  // down
        if (hidx + 1 < history.size()) {
          ++hidx;
          buf = history[hidx];
          cur = buf.size();
        } else {
          hidx = history.size();
          buf.clear();
          cur = 0;
        }
      } else if (key == 'H') {
        cur = 0;
      } else if (key == 'F') {
        cur = buf.size();
      }
      redraw(prompt, buf, cur);
      continue;
    }

    // ---- newline handling ----
    // Return arrives as CR (raw mode disables ICRNL). A bare LF therefore only
    // comes from a paste or from Ctrl-J, so LF always inserts and only CR can
    // submit. CRLF pastes are collapsed via skip_lf.
    if (ch == '\n') {
      if (skip_lf) {
        skip_lf = false;
        continue;
      }
      buf.insert(cur, 1, '\n');
      ++cur;
      out("\r\n");
      continue;
    }
    if (ch == '\r') {
      if (more_pending(kBurstGap)) {  // more bytes behind it => a paste
        buf.insert(cur, 1, '\n');
        ++cur;
        skip_lf = true;  // swallow a following LF
        out("\r\n");
        continue;
      }
      out("\r\n");
      // trailing-whitespace check mirrors Python's `buf.strip()`
      if (buf.find_first_not_of(" \t\r\n") != std::string::npos) {
        history.push_back(buf);
      }
      return {Status::Ok, buf};
    }
    if (ch == '\x03') {  // Ctrl-C
      out("\r\n");
      return {Status::Interrupted, ""};
    }
    if (ch == '\x04') {  // Ctrl-D
      if (buf.empty()) {
        out("\r\n");
        return {Status::Eof, ""};
      }
      continue;
    }
    if (ch == '\x7f' || ch == '\b') {  // backspace
      if (cur > 0) {
        size_t p = prev_cp(buf, cur);
        const bool crossed = buf[p] == '\n';
        buf.erase(p, cur - p);
        cur = p;
        // cannot repaint upward; move up and start the line clean
        if (crossed) out("\x1b[A\r\x1b[K");
      }
      redraw(prompt, buf, cur);
      continue;
    }
    if (ch == '\x15') {  // Ctrl-U: clear the current line
      size_t nl = buf.rfind('\n');
      buf.resize(nl == std::string::npos ? 0 : nl + 1);
      cur = buf.size();
      redraw(prompt, buf, cur);
      continue;
    }
    if (ch == '\x01') {  // Ctrl-A
      cur = 0;
      redraw(prompt, buf, cur);
      continue;
    }
    if (ch == '\x05') {  // Ctrl-E
      cur = buf.size();
      redraw(prompt, buf, cur);
      continue;
    }
    if (b < 0x20 && ch != '\t') continue;  // ignore other controls

    skip_lf = false;
    buf.insert(cur, 1, ch);
    ++cur;
    if (cur == buf.size() && more_pending(0)) {
      // Mid-burst (a paste on a terminal without bracketed paste): append-only
      // echo. A full redraw per character is O(n^2) output and visibly flickers.
      // The last byte of the burst falls through to the real redraw below.
      out(std::string(1, ch));
    } else {
      redraw(prompt, buf, cur);
    }
  }
}

}  // namespace

Line LineReader::read(const std::string& prompt) {
  if (!available()) {  // piped stdin: plain line read
    std::string line;
    if (!std::getline(std::cin, line)) return {Status::Eof, ""};
    return {Status::Ok, line};
  }

  struct termios old {};
  if (tcgetattr(STDIN_FILENO, &old) != 0) return {Status::Eof, ""};
  out(prompt + kPasteOn);

  struct termios raw = old;
  // cfmakeraw, then TCSANOW rather than TCSAFLUSH: TCSAFLUSH DISCARDS pending
  // input, which would silently eat a paste that arrives while the prompt is
  // being drawn, and it blocks until output drains.
  cfmakeraw(&raw);
  raw.c_cc[VMIN] = 1;
  raw.c_cc[VTIME] = 0;
  tcsetattr(STDIN_FILENO, TCSANOW, &raw);

  Line r = loop(prompt, history_);

  tcsetattr(STDIN_FILENO, TCSADRAIN, &old);
  out(kPasteOff);
  return r;
}

}  // namespace lineread
